### ISSUE TITLE
2

### ISSUE THREAD

**Issue by shengzhaotian** [shengzhaotian/firstTestProjiect#5]
`现在我已完整读取所有改动。以下是系统性的评估报告。

---

# 初版代码评估报告

## 总体评价

本 commit 实现了设计文档 **约 70% 的功能面**，覆盖了 21 个文件、+1627 行，整体架构与设计一致。但在 **NPUGraph 兼容性、数值精度验证、测试覆盖** 三个维度存在关键缺陷，部分实现 **无法在生产 full graph 模式下工作**。

---

### 1.1 配置与校验（§3/§11.1） ✅ 基本完整

`validate_npu_selective_hisparse_config` 实现了 30+ 项检查，覆盖了设计 §3 的绝大部分：
- NPU/A5/Decode PD/GLM-5.2/78层/index_topk=2048/ModelSlim/fp8_e4m3
- MTP step=5/draft_tokens=6/eagle_topk∈{None,1}
- DP32/EP32/TP1/CP1/PP1/page_size=1
- graph backend="full"/capture bs 1-8/max_running_requests 显式设置
- `ASCEND_MF_TRANSFER_PROTOCOL=device_urma`

**遗漏**（3 项）：
- 未检查 `disaggregation_decode_enable_radix_cache == False`（设计 §3 第17项）
- 未检查 resolved attention backend == "ascend"（设计 §11.1）
- 未检查 `speculative_num_draft_tokens` 的实际 resolved 值（只检查了 `speculative_eagle_topk`）

### 1.2 候选层推导（§11.2） ✅ 正确

`resolve_dsa_last_shared_layer_ids()` 正确遍历 anchor + 验证后续 `freq-1` 层均为 skip_topk，GLM-5.2 上精确产出 `{5,9,...,77}`。

### 1.3 Pool 物理布局（§11.6） ✅ 核心正确

- `k_buffer` 第一维改为 `resident_main_k_layer_count`（78-n）
- `resident_layer_id_to_slot` 显式映射
- `_get_resident_slot()` 在 `get_kv_buffer`/`get_key_buffer`/`set_kv_buffer` 中统一使用
- selected 层 Host pool 通过 `AscendSelectiveHostKVPool`（memfabric URMA）分配
- `host_sentinel_loc = num_slots - 1`（extra page 最后一行）并初始化为零
- `get_contiguous_buf_infos()` 保持四组 PD ABI，selected MAIN_K 指向 Host HVA
- `get_pd_target_meta()` 生成 per-entry Target 元数据

### 1.4 Prefill CP8 修复（§11.17） ✅ 已合入

- `dsa_npu_indexer.py`: 量化 CP indexer（`npu_quant_lightning_indexer`）已实现
- `ascend_backend.py`: packed FP8 CP SFA（`npu_kv_quant_sparse_flash_attention`）已实现
- `NotImplementedError` 已移除
- RoPE cache init 改为 `layer_id == start_layer`（两处）
- NPU PP 奇偶 rank 收发已实现

### 1.5 Anchor 预取 hook（§11.8） ✅ 位置正确

`forward_dsa_prepare_npu()` 中 Top-K 计算后、return 前插入 `prefetch_for_anchor`，且只在 `is_anchor_layer` 时触发。

### 1.6 Selected 层绕过 MLAProlog（§11.9） ✅

- `mla_preprocess_used` 增加 `and not is_selective_layer`
- `alt_stream q_b_proj` 对 selected 层禁用
- `fused_split_qk_norm` 对 selected 层禁用

### 1.7 Selected 层 attention 分流（§11.10） ✅

`forward_sparse()` 在 `set_kv_buffer` 之后、`get_kv_buffer` 之前分流到 `coordinator.run_selected_attention()`。

### 1.8 MTP verify hook（§11.11） ✅ 位置正确

`run_eagle_verify()` 中 `eagle_sample()` 之后立即调用 `on_verify_result()`。

### 1.9 PD 元数据扩展（§6.6/§11.12/§11.13） ✅ 基本完整

- `pd_extension` 尾帧通过 ZMQ JSON 传输
- `get_pd_target_meta()` 生成完整 Target 元数据表
- `_validate_selective_target_metadata()` 在 PD 连接时做远端校验（4×78 组合顺序、item_lens、memory_kinds、CP layout/sender rank）
- `transfer_engine.py` 支持 `DEVICE_URMA` + A5 gate
- `batch_register` 增加 `strict` 模式

### 1.10 Drain hook（§11.14） ✅

`common.py::release_kv_cache` 和 `schedule_batch.py::release_req` 均增加了 duck-typed drain。

### 1.11 Pool sizing（§11.4） ✅ 公式正确

- `main_num_layers = effective_num_layers - local_selected`
- `_fixed_bias` 包含 staging + metadata + packed_scratch + native_workspace
- `available_bytes - self._fixed_bias` 后再除以 cell_size

---

### 2.1 ❌ NPUGraph 兼容性 — **最高风险**

**问题本质：** Coordinator 使用 Python `torch.npu.Stream/Event` 管理多流，但 NPUGraph capture 无法捕获跨流 Python 对象。

**具体表现：**

`selective_hisparse.py` 中：
```python
self.prefetch_stream = device_module.Stream()  # Python stream
self.backup_stream = device_module.Stream()
...
with device_module.stream(self.prefetch_stream):
    self.host_pool.sparse_copy(...)  # 在 prefetch_stream 上执行
    self.h2d_done.record(self.prefetch_stream)
```

NPUGraph 只捕获 **主 stream 上的 NPU op**。`prefetch_stream` 和 `backup_stream` 上的 `sparse_copy` 和 event record **不会被捕获**，因此 replay 时 H2D/D2H 不会执行。

设计 §10.2/§10.6 明确指出：
> "不能在捕获代码中直接创建新的 Python `torch.npu.Stream/Event`...custom op 必须把 stream/event 生命周期放到扩展侧 handle 中"

并要求新增 `sgl_kernel_npu` custom op（§11.16）：
```
hisparse_prefetch_async / hisparse_prefetch_wait / hisparse_backup / hisparse_drain
```

**当前实现完全没有这些 custom op。** `sgl_kernel_npu` 无任何修改。

**影响：** 验证要求 `decode_graph.backend == "full"`，但实际 full graph 模式下 H2D/D2H 不会执行，selected 层 attention 将读取未填充的 staging（全零），产生错误结果。**Eager 模式可以工作，但生产 graph 模式不行。**

### 2.2 ⚠️ Native Attention 数值正确性 — **需验证**

`compact_packed_attention.py` 的实现逻辑：

```python
latent = (latent_fp8.to(q_nope.dtype) * latent_scale.to(q_nope.dtype))
score = torch.einsum("thd,tkd->thk", q_nope, latent)
score = score + torch.einsum("thr,tkr->thk", q_rope, k_rope)
score = score.to(torch.float32) * scale
```

**潜在问题：**

1. **Scale 乘除方向未验证**：packed record 中的 FP32 scale 是 `npu_dynamic_quant` 产生的 **dequant scale**（`dequant_value = fp8_value * scale`）。代码假设 `latent = fp8 * scale`，但 SFA baseline（`npu_kv_quant_sparse_flash_attention`）内部可能使用 `1/scale` 或不同的乘除顺序。设计 §11.15 明确要求"必须以 `_pack_dsa_fp8_kv_cache()` 及 SFA baseline 的逐字节/数值测试确定"。

2. **V 语义**：代码用同一个 `latent` 同时做 K 和 V（`output = einsum("thk,tkd->thd", prob, latent)`），对应 `attention_mode=2`（MLA 中 V = K_latent）。这在语义上正确，但 SFA 内部可能对 V 有额外的 dequant 处理。

3. **safe-row 实现**：对全无效行，代码用 `slot_zero` trick 保留一个有效 slot 避免 NaN：
```python
safe_mask = valid_mask | ((~valid_query).unsqueeze(-1) & slot_zero.unsqueeze(0))
```
这是合理的，但 `masked_fill` 用 `finfo.min` 而非 `-inf`，softmax 后可能产生微小非零值。最终 `torch.where(valid_query, output, zeros)` 清零了无效行，但有效行中包含 sentinel slot 的注意力权重可能不精确。

4. **Workspace 物化**：`latent` dequant 产生 `[T, K, 512]` BF16，对 `T=48, K=2048` 约 96 MiB；`score` 产生 `[T, H, K]` FP32。这些在 eager 下是临时分配，**graph capture 时会被 graph pool 持有**。`get_native_compact_attention_workspace_spec()` 正确估算了这些，但实际 graph memory 可能超出预算。

### 2.3 ⚠️ Graph 模式下 Python 状态管理

Coordinator 依赖 Python 状态变量：
```python
self._active_selected_layer
self._active_query_tokens
self._published_layer
self._published_packed
```

这些在 capture 时设置，replay 时 **不更新**。对于 graph 模式：
- `_active_query_tokens` 保持 capture-time 值（`capture_bs * 6`）— 这符合设计的静态 shape 要求 ✅
- `_published_packed` 指向 capture-time tensor — graph 会复用同一内存 ✅
- 但 `_active_selected_layer` 的一致性检查在 replay 时使用 stale 值 — 如果多个 selected layer 交错，可能误报

**更关键的是**：`prefetch_for_anchor` 中的 **tensor 计算**（arange、div、indexing 等）在 graph 模式下会被正确捕获和 replay。但 `sparse_copy`（在 prefetch_stream 上）不会被捕获（见 2.1）。

### 2.4 ⚠️ Staging 复用缺少显式 event

设计 §8.4 要求 attention 后记录 `staging_free_event`，下一组 prefetch 等待它。实现用 `prefetch_stream.wait_stream(producer_stream)` 替代，这更保守（等待整个主 stream 而非仅 attention 完成），可能 **减少 overlap 窗口**。

### 2.5 ⚠️ `new_scratch_free_event` 缺失

设计 §11.7.3 要求 `new_packed_scratch` 的复用需要主 stream patch 读取和 backup stream D2H 读取都完成后才能写。实现用 `packed.record_stream(self.backup_stream)` 处理 ownership，但没有显式保证下一次 `publish_new_packed_kv` 写入前两次读取都完成。在多层 selected 场景下可能产生 **scratch 竞态**。

### 2.6 ⚠️ `h2d_count` 语义

```python
self.h2d_count.copy_(self.num_real_reqs * (SELECTIVE_VERIFY_TOKENS * SELECTIVE_TOPK))
```

`num_real_reqs` 在 `model_runner.py` 中设为 `forward_batch.batch_size`。但在 DP attention 场景下，`batch_size` 可能包含 padding 请求。设计 §6.4 要求 `B_local = block_tables.shape[0]`（实际 per-rank 请求数）。如果 `batch_size > block_tables.shape[0]`，会多传无效记录。

### 2.7 ⚠️ `ForwardBatch` 未传播 coordinator

设计 §11.3 要求在 `ForwardBatch` 上新增 `npu_selective_hisparse_coordinator` 和 `npu_selective_graph_mode` 字段。实现改为通过 `token_to_kv_pool.selective_coordinator` 访问。这简化了代码但：
- 无法区分 eager/graph 模式（`npu_selective_graph_mode` 缺失）
- `forward_batch_info.py` 完全未修改

---

## 三、缺失的部分

| 缺失项 | 设计章节 | 影响 |
|---|---|---|
| **`sgl_kernel_npu` custom op** | §11.16 | **阻塞性** — graph 模式不可用 |
| **任何测试** | §14 | **阻塞性** — 无正确性验证 |
| `ForwardBatch` 字段传播 | §11.3 | 中 — graph/eager 分流不明确 |
| `disaggregation_decode_enable_radix_cache` 校验 | §3 第17项 | 低 — PD hook 通常自动设置 |
| Resolved attention backend 校验 | §11.1 | 低 |
| `staging_free_event` 显式管理 | §8.4 | 中 — overlap 窗口缩小 |
| `new_scratch_free_event` | §11.7.3 | 中 — 多层 scratch 竞态风险 |
| `verify_generation` 实际使用 | §8.10 | 低 — 当前只递增不使用 |
| `device_assert_no_true()` NPU 异步断言 | §11.7.1 | 中 — 当前用 Python `RuntimeError`，graph 下不执行 |
| Prefill 端 `pd_extension` 发布 | §11.12 | 中 — Prefill 未发送 component_kind/layer_id |
| Mixed CPU snapshot retraction 完整实现 | §8.8 | 中 — `get_cpu_copy` 改了但未验证 |

---

## 四、按设计阶段的可工作性判定

| 阶段 | 状态 | 说明 |
|---|---|---|
| **A: 外部 Gate + Native POC** | ⚠️ **部分** | Prefill CP8 代码已合入但无测试；memfabric H2D 可工作；native attention 逻辑正确但数值未验证 |
| **B: 配置/Pool/PD** | ✅ **基本完成** | Pool 布局、PD 元数据、sizing 均实现；缺少少量校验项 |
| **C: 单层 eager 闭环** | ⚠️ **可能可工作** | eager 模式下流程完整，但需 NPU 实机验证 native attention 数值和 `sparse_copy` 行为 |
| **D: NPUGraph 闭环** | ❌ **不可工作** | 缺少 custom op，多流操作无法被 graph 捕获 |
| **E: 多层+PD E2E** | ❌ **阻塞** | 依赖 D 完成 |
| **F: 性能** | ❌ **阻塞** | 依赖 D/E 完成 |

---

### 能工作（eager 开发对齐模式）
在 **关闭 NPUGraph**（或 `decode_graph.backend != "full"`）的情况下，本实现的 eager 路径 **理论上可以跑通单层或多层 selective HiSparse**。前提是：
1. `memfabric_hybrid.offload.sparse_copy` 能正确执行随机 gather
2. Native attention 的 656B 解包和 dequant 数值与 SFA baseline 对齐
3. NPU multi-stream 在 eager 下正确 overlap

### 不能工作（生产 full graph 模式）
验证强制要求 `decode_graph.backend == "full"`，但实现使用 Python Stream/Event 管理多流，**无法被 NPUGraph 捕获**。这是设计 §10.6 预判的"硬前置"问题——需要 `sgl_kernel_npu` custom op 把 stream/event 生命周期封装到 C++ 扩展中。

### 三个必须解决的阻塞项

| 优先级 | 阻塞项 | 解决方案 |
|---|---|---|
| **P0** | `sgl_kernel_npu` custom op 缺失 | 新增 `hisparse_prefetch_async/wait/backup/drain`，在 C++ 内部管理 stream/event |
| **P0** | 测试完全缺失 | 至少补充 §14.1 的 CPU unit tests 和 §14.2 的 NPU micro-tests |
| **P1** | Native attention 数值未验证 | 与 SFA baseline 逐字节比较 656B record 解包、dequant scale 方向、softmax 容差 |

### 建议的下一步

1. **先验证 eager 正确性**：临时关闭 graph 校验，在 NPU 上跑 `[5]` 单层，比较 logits/accept_lens 与 baseline
2. **做 NPUGraph POC**：按设计 §10.6 建议，先做单 group 的图内 Host gather capture/replay POC
3. **补充 native attention 数值测试**：固定输入，比较 `compact_packed_mla_attention` 与 `npu_kv_quant_sparse_flash_attention` 的输出`
